### PR TITLE
feat(new sink): Initial `aws_cloudwatch_metrics` sink implementation 

### LINKED ISSUE #660: New `aws_cloudwatch_metrics` sink
We'd like to extract metrics from logs using `log_to_metric` and ship them to CloudWatch metrics as a sink.

### PR DESCRIPTION
**Pull request by loony-bean**
Closes #660 

This PR adds basic support for sending CloudWatch metrics via [API](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricData.html).

It allows to send Counter, Gauge (only absolute ones) and Histogram type of metrics in a mini-batched fashion to CloudWatch endpoint.

This is our second 'real' metrics sink, and adding it gives us a new useful perspective on our metrics type system and flow. Below is a list of questions I think we need to consider before moving on with this integration.

1. Event timestamps

From API perspective It is possible to send to CloudWatch the real (e.g. parsed from the incoming logs) timestamps of the events. However we store timestamps in implicit log data that is gone during `log_to_metric` transformation. It might not be a big deal, because CW will apply it's own receive time timestamp to event by default, but still it's a limitation. Possible solution could be to add `timestamp` to our metric model, and populate that field in `log_to_metric`.

2. Relative gauges and sets

Seems like the majority of metric sinks will need us to implement some sort of state to enable reporting of relative gauges and such. I'm not sure if this 'accumulation' should be implemented in each sink separately, or there might be a better design solution to this?

**Update**: Timestamps support is in place, and relative sets are also supported. `Sets` however will need some design, and will be addressed in a follow up issue.

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice, excited to get this merged. @lukesteensen is on vacation this week so I'll answer your questions along with @LucioFranco .

1. Good catch. I think we should add a timestamp field to our metrics to preserve the original time, if present.
2. I think it's fine to implement this within each individual sink for now. There's always going to be shared behavior across sinks, and as long as we extract the code in a reusable fashion it should not be a proble. For example, #678 introduced a method to unflatten log events, which is used throughout all log sinks. Finally, as we progress the patterns will become more apparent, and I fully expect us to take a step back and make design changes once we have those. So, do what you need to do in the CW metrics sink for now.

Let me know if that helps.

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+                val: i as f32,
+                sample_rate: 100,
+            });
+            events.push(event);
+        }
+
+        let stream = stream::iter_ok(events.clone().into_iter());
+
+        let pump = sink.send_all(stream);
+        rt.block_on(pump).unwrap();
```
Comment (LucioFranco) [inline src/sinks/aws_cloudwatch_metrics.rs]: is there any way we can check that these items were inserted correctly?

**Comment by loony-bean** [comment]
Thanks @LucioFranco for all the comments! Really appreciate such attention to details, I needed it here. I will address the comments in place.

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    ) -> Result<super::RouterSink, String> {
+        let client = Self::create_client(config.region.clone().try_into()?)?;
+
+        let batch_size = config.batch_size.unwrap_or(5);
+        let batch_timeout = config.batch_timeout.unwrap_or(1);
+
+        let timeout = config.request_timeout_secs.unwrap_or(30);
+        let in_flight_limit = config.request_in_flight_limit.unwrap_or(5);
+        let rate_limit_duration = config.request_rate_limit_duration_secs.unwrap_or(1);
+        let rate_limit_num = config.request_rate_limit_num.unwrap_or(5);
```
[the remark this answers is not in the document]
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I'm thinking about bumping it to 25, what do you think? I'm a bit afraid of making Vector a 'busy neighbour' for possible other applications publishing metrics into CW, assuming this limit is applied on account level.

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+impl CloudWatchMetricsService {
+    pub fn new(
+        config: CloudWatchMetricsSinkConfig,
+        acker: Acker,
+    ) -> Result<super::RouterSink, String> {
+        let client = Self::create_client(config.region.clone().try_into()?)?;
+
+        let batch_size = config.batch_size.unwrap_or(5);
```
[the remark this answers is not in the document]
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I don't thick there is a formula for this. However we are not using bytes as batching unit, but Events count. Each event is put into `PutMetricData` as separate `MetricDatum`. The limit for `MetricDatum` count per call is 20, and I think this could be a reasonable default for batch_size.

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            Ok(CloudWatchClient::new(region))
+        }
+    }
+}
+
+use bytes::Bytes;
+
+impl Partition<Bytes> for Event {
+    fn partition(&self) -> Bytes {
+        "key".into()
```
[the remark this answers is not in the document]
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I was thinking about partitioning for some time, and I didn't come with any better scheme to use. Maybe we should open a follow up issue to investigate further payload structure optimisations.

Namespaces are currently configured at Sink level (not sure if that's what users will find useful /cc @binarylogic), thus they will not work as partition keys.

Partitioning by metric name will not bring much value, because:
a) in some cases we will need to pass timestamps, and timestamp is part of `MetricDatum`, and timestamped datums cannot really be compressed
b) labels aka 'dimensions' are considered part of metric name, and those could have some extra variability, and also are part of `MetricDatum`.

To summarise, seems like the only simple batching strategy we can implement is to pack 20 `MetricDatum` objects into `PutMetricData` request, and in each `MetricDatum` we will not benefit from any kind of partitioning. Hope this all makes sense!

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 1 above)
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I've tried to find a working combination of localstack/moto, rusoto and friends to make this work, but I ran out of time unfortunately.

- list_metrics was not working out of the box
- get_metric_statistics either
- get_metric_data is not yet implemented https://github.com/spulec/moto/issues/2239

Not sure if I should skip this part of test for now, or spend some more time to make it work.

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/sinks/aws_cloudwatch_metrics.rs]: Interesting, point, I do think though that this is configurable and we should probably default to the most that cloudwatch metrics supports. Then document that if you are using other applications and putting then you should lower it? What do you think?

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/sinks/aws_cloudwatch_metrics.rs]: I think maybe partitioning shouldn't be in the scope of this PR then. I think we could use some more discussion around that. In that case, this PR should be updated to reflect that. That means instead the sink combinator should be changed to just batched_with_min.

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 1 above)
Comment (LucioFranco) [reply src/sinks/aws_cloudwatch_metrics.rs]: So pretty much there is no way to fetch metrics? 😢

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I agree with that. By the way, what should we do in case user is trying to provide configuration values beyond limits (especially those that cannot be changed)? Or we should leave it between user and AWS.

[the remark this answers is not in the document]
Comment (LucioFranco) [reply src/sinks/aws_cloudwatch_metrics.rs]: I think in that case, we let them do it. They will get errors about being throttled so they should then be able to adjust.

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: Omg how I just overcomplicated it :)

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 1 above)
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I haven't found it yet 🤷‍♂

Comment (LucioFranco) [reply src/sinks/aws_cloudwatch_metrics.rs]: Time to open an angry issue with localstack 😂 

Anyways that is fine, though ideally we should be checking but I'm not sure how to accomplish this without us making a PR to localstack.

**Comment by binarylogic** [comment]
@loony-bean now that https://github.com/timberio/vector/pull/726 is merged we should probably implement it here. Nice call on that 👍 .

**Comment by loony-bean** [comment]
@LucioFranco Could we have another round of review? I believe that the comments from the previous one are resolved.

One thing that is intentionally left outside of scope is `Sets` support. I'm going to create a follow up issue for it.

**Comment by binarylogic** [comment]
> One thing that is intentionally left outside of scope is Sets support. I'm going to create a follow up issue for it.

Sounds good, can you open the issue before we merge?

**Comment by loony-bean** [comment]
> Sounds good, can you open the issue before we merge?

Yep, just opened #747

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 5]
```diff
@@ -1,28 +1,35 @@
...
 use rusoto_core::{Region, RusotoFuture};
 use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
 use std::{convert::TryInto, time::Duration};
 use tower::{Service, ServiceBuilder};
-use tracing::field;
 use tracing_futures::{Instrument, Instrumented};
 
+#[derive(Clone, Default)]
+struct CloudWatchMetricsState {
```
Comment (LucioFranco) [inline src/sinks/aws_cloudwatch_metrics.rs]: very minor nit but this name is getting a bit long, should we think about maybe just renaming it to `State` since its just an internal private struct?

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 6]
```diff
@@ -156,9 +163,9 @@ impl Service<Vec<Event>> for CloudWatchMetricsService {
 
     fn call(&mut self, items: Vec<Event>) -> Self::Future {
         let namespace = self.config.namespace.clone();
-        let input = encode_events(items, namespace).unwrap();
+        let input = encode_events(&mut self.state, items, namespace).unwrap();
```
Comment (LucioFranco) [inline src/sinks/aws_cloudwatch_metrics.rs]: maybe makes sense to make this apart of `CloudWatchMetricsService`?

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 7]
```diff
@@ -1,28 +1,35 @@
...
-use tracing::field;
 use tracing_futures::{Instrument, Instrumented};
 
+#[derive(Clone, Default)]
+struct CloudWatchMetricsState {
+    gauges: HashMap<String, f64>,
+}
+
 #[derive(Clone)]
 pub struct CloudWatchMetricsService {
```
Comment (LucioFranco) [inline src/sinks/aws_cloudwatch_metrics.rs]: take it or leave it: but in cloudwatch logs we usually name the struct that implements service CloudwatchLogsSvc instead of Service to indicate that it is just an implementation not a trait. I'm fine with keeping it how it is now, just wanted to make sure you were aware.

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 8]
```diff
@@ -177,38 +184,80 @@ impl RetryLogic for CloudWatchMetricsRetryLogic {
...
+            {
+                true
+            }
             _ => false,
         }
     }
 }
 
-fn encode_events(events: Vec<Event>, namespace: String) -> Result<PutMetricDataInput, ()> {
+fn timestamp_to_string(timestamp: DateTime<Utc>) -> String {
```
Comment (LucioFranco) [inline src/sinks/aws_cloudwatch_metrics.rs]: We could make this fn public https://github.com/timberio/vector/blob/master/src/event/mod.rs#L281 and use that one instead of redefining it?

File: src/sinks/aws_cloudwatch_metrics.rs [hunk 9]
```diff
@@ -177,38 +184,80 @@ impl RetryLogic for CloudWatchMetricsRetryLogic {
...
+    timestamp.to_rfc3339_opts(SecondsFormat::Millis, true)
+}
+
+fn encode_events(
+    state: &mut CloudWatchMetricsState,
+    events: Vec<Event>,
+    namespace: String,
+) -> Result<PutMetricDataInput, ()> {
```
Comment (LucioFranco) [inline src/sinks/aws_cloudwatch_metrics.rs]: does an option make more sense here than a result since there is no error type?

**Comment by lukesteensen** [comment]
This is looking really good, great work @loony-bean!

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 9 above)
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: You are right. Seems like even plain `PutMetricDataInput` will do.

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 8 above)
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: I'm not sure, because there is a difference in SecondsFormat: AutoSi vs Millis.

The CloudWatch spec for timestamp is vague. It contains a knows issue of documenting this field such as "expressed as the number of milliseconds since Jan 1, 1970 00:00:00 UTC", however it's not really true. The examples and tooling around this API all seems to use ms precision, and I've decided to do the same to feel safer.

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 7 above)
Comment (loony-bean) [reply src/sinks/aws_cloudwatch_metrics.rs]: Thanks!

File: src/sinks/aws_cloudwatch_metrics.rs (hunk 8 above)
Comment (LucioFranco) [reply src/sinks/aws_cloudwatch_metrics.rs]: Ok that is fine, I probably didn't notice that detail 😄
